test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs

```diff
@@ -0,0 +1,647 @@
+        {
+            AuthenticationScheme = scheme ?? OpenIdConnectAuthenticationDefaults.AuthenticationScheme;
+        }
+
+        public string AuthenticationScheme
+        {
+            get;
+            set;
+        }
+
+        public void Authenticated(ClaimsPrincipal principal, IDictionary<string, string> properties, IDictionary<string, object> description)
+        {
+        }
+
+        public void NotAuthenticated()
+        {
+        }
+    }
+
+    /// <summary>
+    /// This type is used to call 'protected' methods on OpenIdConnectAuthenticationHandler.
+    /// Makes writing comprehensive tests easier
+    /// </summary>
+    public class OpenIdConnectAuthenticationHandlerPublic : OpenIdConnectAuthenticationHandler
+    {
+        public OpenIdConnectAuthenticationHandlerPublic(ILogger logger)
+            : base(logger)
+        {
+        }
+
+        public async Task BaseInitializeAsyncPublic(AuthenticationOptions options, HttpContext context)
+        {
+            await base.BaseInitializeAsync(options, context);
+        }
+
+        public override bool ShouldHandleScheme(string authenticationScheme)
+        {
+            return true;
+        }
+
+        public override void Challenge(IChallengeContext context)
+        {
+            return;
+        }
+
+        protected override void ApplyResponseChallenge()
+        {
+            return;
+        }
+
+        protected override async Task ApplyResponseChallengeAsync()
+        {
+            var redirectToIdentityProviderNotification = new RedirectToIdentityProviderNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions>(Context, Options)
+            {
+            };
+
+            await Options.Notifications.RedirectToIdentityProvider(redirectToIdentityProviderNotification);
+        }
+    }
+
+    /// <summary>
+    /// This type is used to call 'protected' methods on OpenIdConnectAuthenticationMiddleware.
+    /// Makes writing comprehensive tests easier
+    /// </summary>
+    public class OpenIdConnectAuthenticationMiddlewarePublic : OpenIdConnectAuthenticationMiddleware
+    {
+        public static IDataProtectionProvider DataProtectionProvider;
+        public static OpenIdConnectAuthenticationMiddlewarePublic THIS;
+
+        public OpenIdConnectAuthenticationMiddlewarePublic(
+            RequestDelegate next,
+            IDataProtectionProvider dataProtectionProvider,
+            ILoggerFactory loggerFactory,
+            IOptions<OpenIdConnectAuthenticationOptions> options,
+            ConfigureOptions<OpenIdConnectAuthenticationOptions> configureOptions
+            )
+        : base(next, dataProtectionProvider, loggerFactory, options, configureOptions)
+        {
+            DataProtectionProvider = dataProtectionProvider;
+            OptionsPublic = options.Options;
+            SecureDataFormat = Options.StateDataFormat;
+            THIS = this;
```

Comment:
Hmm some of this test code looks suspicious to me. We can't have singletons in tests because tests run in parallel.